1.7.0 documentation (#360)
IBM/ibm-block-csi-driver: docs/content/configuration/csi_ug_config_create_secret_topology.md

# Creating a Secret with topology awareness

Create an array secret YAML file to define the storage credentials (username and password) and address. Use this information for creating a Secret that is topology aware.

**Note:** If you are not using the CSI Topology feature, follow the steps in [Creating a Secret](csi_ug_config_create_secret.md).

> Within the Secret, each management ID (here, represented by `demo-management-id-x`), is used to identify the storage system within other configuration files.

Reviewer: also, @ArbelNathan should we add any limitation to the value of management id (length, charset, etc...)?

> Within the Secret, each management ID (here, represented by `demo-management-id-x`), is used to identify the storage system within other configuration files.

Reviewer 2: amm the only limitation we have for it is to answer to the regex we defined (`'^[a-zA-Z0-9][a-zA-Z0-9-_.]*[a-zA-Z0-9]$'`)
its meaning is we can start and end with characters and numbers only and in the middle have only characters, numbers and the symbols `-`, `_`, and `.`.

> Within the Secret, each management ID (here, represented by `demo-management-id-x`), is used to identify the storage system within other configuration files.

Reviewer: thanks! FYI @rivkap

> Within the Secret, each management ID (here, represented by `demo-management-id-x`), is used to identify the storage system within other configuration files.

Author: DONE

# Creating a Secret with topology awareness

Create an array secret YAML file in order to define the storage credentials (username and password) and address. Use this information for creating a Secret that is topology aware.

**Note:** If you are not using the CSI Topology feature, follow the steps in [Creating a Secret](csi_ug_config_create_secret.md).

> Use the Secret file to connect a worker node to a storage system.

Reviewer 2: that can be sound like by using that secret we are automatically connecting the actual worker to the storage..

Reviewer: ```suggestion
Use the Secret file to allow connecting to a storage system.
```

Reviewer 2: isn't it the purpose of every secret? 
maybe something more like: 

> Use the Secret file to reflect the connection mapping of the worker nodes to a storage system.

what do you think?

Author: @ArbelNathan  @oriyarde  I am actually not sure why we have this sentence. it's not in the "original" secret file. Can I just remove it all together? or is it needed for topology and I should write it as was suggested:
"Use the Secret file to reflect the connection mapping of the worker nodes to a storage system."

Reviewer: I would remove it altogether

1. Create the secret file, similar to the following demo-secret-config.json:

    The `management_address` field can contain more than one address, with each value separated by a comma.

    ```
        {

>      "demo-system-id-1": {

Reviewer 2: please align with example yamls

Author: done

       "username": "demo-username-1",
       "password": "demo-password-1",
        "management_address": "demo-management-address-1",
        "supported_topologies": [
         {
           "topology.block.csi.ibm.com/demo-region": "demo-region-1",
           "topology.block.csi.ibm.com/demo-zone": "demo-zone-1"
         }
       ]
     },
     "demo-system-id-2": {
       "username": "demo-username-2",
       "password": "demo-password-2",
       "management_address": "demo-management-address-2",
       "supported_topologies": [
         {
           "topology.block.csi.ibm.com/demo-region": "demo-region-2",
           "topology.block.csi.ibm.com/demo-zone": "demo-zone-2"
         }
       ]
     }
   }
     ```
       
2. Apply the secret using the following command:

    `kubectl create secret generic <secret name> -n <secret namespace> --from-file=config=demo-secret-config.json`
    

>      The `secret/<secret name> created` message is emitted.

Reviewer 3: I think a hyphen is missing here

Author: @lisyukanna  does the name have to be one word / hyphenated? or can it be more than that? if yes, i'll update in all relevant places

Reviewer 3: as far as I know, yes it should be one word, @oriyarde ?

Reviewer: don't know about `have to`, but that is the convention I know.
and regardless, I prefer the actual name, e.g. `demo-secret`

Author: I updated a few emitted messages so that there was more consistency.
I also updated here the secret-name with a hyphen and added the same to the command above as well.
DONE

# Creating a Secret with topology awareness

Create an array secret YAML file in order to define the storage credentials (username and password) and address. Use this information for creating a Secret that is topology aware.

**Note:** If you are not using the CSI Topology feature, follow the steps in [Creating a Secret](csi_ug_config_create_secret.md).

Use the Secret file to connect a worker node to a storage system.

> Within the Secret, each `management-id` (here, represented by `demo-management-id-x`), is used to identify the storage system within other configuration files.

Reviewer: if there was a name for this field, our parameters naming convention actually has underscores (and not hyphens):
```suggestion
Within the Secret, each `management_id` (here, represented by `demo-management-id-x`), is used to identify the storage system within other configuration files.
```
hopefully this would be easier to relate to `by_management_id`

Author: @oriyarde  

1. If the convention is to use underscores, then shouldn't `demo-management-id-x` be changed in the example json as well? This is where I took the information from.
2. Upon further inspection of the JSON, perhaps I should write: "Within the Secret, each **`management_address`** (here, represented by xxxx)..."

FYI @ArbelNathan

Reviewer: 1. we have a different convention for field names and field values. the naming convention comes from python variables snake_case, and IMO the hyphenated value convention helps distinguish it from the name.
2. `management_address` is a different field than `management_id`

Author: I see now - 
in any case - as you can see in my markup below - the management-id that I'm referring to (with the clear example) is for the name which, in the json example is also with the hyphens. 

![image](https://user-images.githubusercontent.com/54570654/134345358-ae1c5392-7baa-450e-8c18-5a60c312d4f3.png)

Reviewer: I see, then I suggest:
```suggestion
Within the Secret, each management id (here, represented by `demo-management-id-x`), is used to identify the storage system within other configuration files.
```
since there is no hyphenated `management-id` without `demo-` anywhere

Author: DONE

> 1. Create the secret file, similar to the following demo-secret-config.json:

Author: DONE

    The `management_address` field can contain more than one address, with each value separated by a comma.

    ```
      {
    "demo-management-id-1": {
      "username": "demo-username-1",
      "password": "demo-password-1",
      "management_address": "demo-management-address-1",
      "supported_topologies": [
        {
          "topology.block.csi.ibm.com/demo-region": "demo-region-1",
          "topology.block.csi.ibm.com/demo-zone": "demo-zone-1"
        }
      ]
    },
    "demo-management-id-2": {
      "username": "demo-username-2",
      "password": "demo-password-2",
      "management_address": "demo-management-address-2",
      "supported_topologies": [
        {
          "topology.block.csi.ibm.com/demo-region": "demo-region-2",
          "topology.block.csi.ibm.com/demo-zone": "demo-zone-2"
        }
      ]
    }
  }

>      ```

Reviewer: this appears to be outside the block formatting

Author: DONE

Use one of the following procedures to create and apply the secret:

## Creating an array secret file that is Topology Aware
1. Create the secret file, similar to the following demo-secret-config.json:

    The `management_address` field can contain more than one address, with each value separated by a comma.

    ```
        {
     "demo-system-id-1": {
       "username": "demo-username-1",
       "password": "demo-password-1",
        "management_address": "demo-management-address-1",
        "supported_topologies": [
         {
           "topology.block.csi.ibm.com/demo-region": "demo-region-1",
           "topology.block.csi.ibm.com/demo-zone": "demo-zone-1"
         }
       ]
     },
     "demo-system-id-2": {
       "username": "demo-username-2",
       "password": "demo-password-2",
       "management_address": "demo-management-address-2",
       "supported_topologies": [
         {
           "topology.block.csi.ibm.com/demo-region": "demo-region-2",
           "topology.block.csi.ibm.com/demo-zone": "demo-zone-2"
         }
       ]
     }
   }
     ```
       
2. Apply the secret using the following command:

>     `kubectl apply -f demo-secret-config.json`

Author: @ArbelNathan  please resolve

    

     The `secret/<NAME> created` message is emitted.

## Creating an array secret via command line
**Note:** This procedure is applicable for both Kubernetes and Red Hat® OpenShift®. For Red Hat OpenShift, replace `kubectl` with `oc` in all relevant commands.

Create the secret using the following command:

 ```
 kubectl create secret generic <NAME> --from-literal=username=<USER> --from-literal=password=<PASSWORD>--from-literal=management_address=<ARRAY_MGMT> -n <namespace>

>  ```

Reviewer 2: it's not relevant for Topology

Author: @ArbelNathan  please resolve
